Refactor strtree to not use lgeos / ctypes

jorisvandenbossche opened the pull request:
This is for the shapely-2.0 branch, to do the equivalent of https://github.com/Toblerity/Shapely/pull/1112 or https://github.com/Toblerity/Shapely/pull/1064 (and ideally only gets done when one of those PRs is merged).

This shows how the `lib.STRtree` (the C object) will just contain primary keys, which are mapped to geometries (or other items) in the `query` method using the returned indices.

snorfalorpagus commented on shapely/strtree.py:
```diff
@@ -168,18 +147,17 @@ def query(self, geom):
 
-        return result
+    def query_items(self, geom):
+        """
```
Missing docstring

snorfalorpagus commented on shapely/strtree.py:
```diff
@@ -168,18 +147,17 @@ def query(self, geom):
+        return self.geometries[indices]
 
-        return result
+    def query_items(self, geom):
```
This function needs a test

snorfalorpagus commented on shapely/strtree.py:
```diff
@@ -214,23 +192,26 @@ def nearest(self, geom):
+        return self.geometries[idx]
+
+    def nearest_item(self, geom):
+        """
```
Missing docstring

snorfalorpagus commented on shapely/strtree.py:
```diff
@@ -214,23 +192,26 @@ def nearest(self, geom):
+        idx = indices[1, 0]
+        return self.geometries[idx]
+
+    def nearest_item(self, geom):
```
This function needs a test

snorfalorpagus commented on shapely/strtree.py:
```diff
@@ -75,44 +72,26 @@ class STRtree:
+        else:
+            return (STRtree, (self.geometries, ))
 
     def query(self, geom):
```
This function now returns an `np.array` instead of a list which is causing the doctest to fail.

mwtoews replied:
docstring currently has "array or list of geometry objects.". Perhaps it should be updated to say just "array".

jorisvandenbossche replied:
Ah, yes, updated the docstrings.

snorfalorpagus reviewed (changes requested):
Added some minor comments.

I like the new `query_items` function. This is a real pain in the current API.

jorisvandenbossche commented:
@snorfalorpagus 

> I like the new `query_items` function. This is a real pain in the current API.

That's actually already introduced in the main branch (and so will be in shapely 1.8), coming from #1112 / #1064

Related to the missing docstrings and tests, they were also introduced in #1112, so I need to rebase this branch to pull those in (I made this branch with just similar code changes before #1112 was merged).

jorisvandenbossche commented:
I updated the shapely-2.0 branch with the changes of latest master, and so also updated this branch (so docstrings and tests are now already included)

jorisvandenbossche commented:
Note that this PR disables the `exclusive` keyword again in the `nearest_item`/`nearest_geom` methods (which was introduced in https://github.com/Toblerity/Shapely/pull/1166), since pygeos does not yet support this. 
Of course ideally we fix this before an actual 2.0 release, but I would prefer to not keep up this PR until it's supported in the pygeos version.

mwtoews reviewed (commented):
Admittedly, I don't know STRTree very well, but I've tested this locally and it seems to be in working order.

Perhaps the docstrings could be updated to say if an array is returned from (e.g.) `query`. Also, should `query` and `nearest` be at least deprecated? (These are still in consideration to be removed for 2.0)

Is there an issue in pygeos for the `exclusive` keyword? It hopefully shouldn't be too difficult to implement.

jorisvandenbossche commented:
> Is there an issue in pygeos for the exclusive keyword? It hopefully shouldn't be too difficult to implement.

Didn't do that yet (will do now). Somewhat related, I did just open an issue about some other API incompatibilities between the pygeos and Shapely STRtree: https://github.com/pygeos/pygeos/issues/398

jorisvandenbossche commented:
Opened https://github.com/pygeos/pygeos/issues/400 for the `exclusive` keyword.

jorisvandenbossche commented:
> Also, should `query` and `nearest` be at least deprecated? (These are still in consideration to be removed for 2.0)

Not fully sure, what I think that's a separate issue to discuss.

jorisvandenbossche commented:
Since this is only replacing the internals of the tree with pygeos (and not changing any API), I am going to merge this, so I can update https://github.com/Toblerity/Shapely/pull/1163
